src/codec.rs

```diff
@@ -664,19 +696,51 @@ impl Decode for Compact<u64> {
 		let prefix = input.read_byte()?;
 		Ok(Compact(match prefix % 4 {
 			0 => prefix as u64 >> 2,
-			1 => u16::decode(&mut PrefixInput{prefix: Some(prefix), input})? as u64 >> 2,
-			2 => u32::decode(&mut PrefixInput{prefix: Some(prefix), input})? as u64 >> 2,
+			1 => {
+				let x = u16::decode(&mut PrefixInput{prefix: Some(prefix), input})? >> 2;
+				if x > 0b00111111 && x <= 0b00111111_11111111 {
+					x as u64
+				} else {
+					return Err(U64_OUT_OF_RANGE.into());
+				}
+			},
+			2 => {
+				let x = u32::decode(&mut PrefixInput{prefix: Some(prefix), input})? >> 2;
+				if x > 0b00111111_11111111 && x <= u32::max_value() >> 2 {
+					x as u64
+				} else {
+					return Err(U64_OUT_OF_RANGE.into());
+				}
+			},
 			3|_ => match (prefix >> 2) + 4 {
-				4 => u32::decode(input)? as u64,
-				8 => u64::decode(input)?,
+				4 => {
+					let x = u32::decode(input)?;
+					if x > u32::max_value() >> 2 {
```

Comment:
This is also incorrect, we store the "length" information in an extra byte in front of the actual value.